src/java/org/apache/cassandra/db/compaction/ShardManagerNodeAware.java

```diff
@@ -0,0 +1,280 @@
+    {
+        var splitPointCount = shardCount - 1;
+        // TODO is it safe to get tokens here and then endpoints later without synchronization?
+        var sortedTokens = tokenMetadata.sortedTokens();
+        if (splitPointCount > sortedTokens.size())
+        {
+            // Need to allocate tokens within node boundaries.
+            var endpoints = tokenMetadata.getAllEndpoints();
+            double addititionalSplits = splitPointCount - sortedTokens.size();
+            var splitPointsPerNode = (int) Math.ceil(addititionalSplits / endpoints.size());
+            // Compute additional tokens since we don't have enough
+            for (var endpoint : endpoints)
+                sortedTokens.addAll(TokenAllocation.allocateTokens(tokenMetadata, rs, endpoint, splitPointsPerNode));
+            // Sort again since we added new tokens.
+            sortedTokens.sort(Token::compareTo);
+        }
+        var splitPoints = findTokenAlignedSplitPoints(sortedTokens.toArray(TOKENS), shardCount);
+        return new NodeAlignedShardTracker(shardCount, splitPoints);
+    }
+
+    private Token[] findTokenAlignedSplitPoints(Token[] sortedTokens, int shardCount)
+    {
+        // Short circuit on equal
+        if (sortedTokens.length == shardCount - 1)
+            return sortedTokens;
+        var evenSplitPoints = computeUniformSplitPoints(tokenMetadata.partitioner, shardCount);
+        var nodeAlignedSplitPoints = new Token[shardCount - 1];
+
+        // UCS requires that the splitting points for a given density are also splitting points for
+        // all higher densities, so we pick from among the existing tokens.
+        int pos = 0;
+        for (int i = 0; i < evenSplitPoints.length; i++) {
+            Token value = evenSplitPoints[i];
+            pos = Arrays.binarySearch(sortedTokens, pos, evenSplitPoints.length, value);
+
+            if (pos >= 0)
+            {
+                // Exact match found
+                nodeAlignedSplitPoints[i] = sortedTokens[pos];
+                pos++;
+            }
+            else
+            {
+                // pos is -(insertion point) - 1, so calculate the insertion point
+                pos = -pos - 1;
+
+                // Check the neighbors
+                Token leftNeighbor = sortedTokens[pos - 1];
+                Token rightNeighbor = sortedTokens[pos];
+
+                // Choose the nearest neighbor. By convention, prefer left if value is midpoint.
+                if (value.size(leftNeighbor) <= value.size(rightNeighbor))
+                {
+                    nodeAlignedSplitPoints[i] = leftNeighbor;
+                    // No need to bump pos because we decremented it to find the right split token.
+                }
+                else
+                {
+                    nodeAlignedSplitPoints[i] = rightNeighbor;
+                    pos++;
+                }
+            }
+        }
+
+        return nodeAlignedSplitPoints;
```

Comment:
I am pretty sure I already did this, though it might not be in the ideal way. As we iterate over `sortedTokens`, we find the position of the nearest neighbor, and then we make set `pos` to the index of the next token in `nodeAlignedSplitPoints`. We then use `pos` as the lower bound in the binary search to find the closest split point. I haven't confirmed that this solution maintains the rule that `UCS requires that the splitting points for a given density are also splitting points for all higher densities`.

I can see that the recursive solution would trivially ensure `UCS requires that the splitting points for a given density are also splitting points for all higher densities`. Do you think we should prefer that approach?

Reply:
I think that mine works assuming we have the power-of-two token allocator.

Reply:
Yes, I think this could work too (if we don't select a specific token, it would be because we selected it for a different split point). Added a couple of comments on ensuring no repetition, and not exhausting the sorted tokens too early.

This method needs to be unit-tested.

Reply:
I hit a bug in this part of the code when working on adding unit tests and I think that it would be better to have a cleaner and clearer design. In general, I don't yet see how we guarantee that we select the same tokens at successively higher levels as tokens are added and removed to the sorted tokens list. Is it safe to assume a certain 

> One thing we can do is remove entries from the node-aligned list when we use them, and enforce that the ones picked for smaller shard counts are also picked by recursively generating for `shardCount / 2` first until `shardCount` is not divisible by 2.

I tried implementing this, but I ran into trouble because I assumed I would have enough split points if I broke the space up by the nearest token to the `midpoint`:

```java
    private void findTokenAlignedSplitPoints(Token[] sortedTokens, int min, int max, Token left, Token right, int splitPointCount, List<Token> splitPoints)
    {
        splitPointCount--;
        var midpoint = partitioner.midpoint(left, right);
        var index = Arrays.binarySearch(sortedTokens, min, max, midpoint);
        if (index < 0)
        {
            // -(insertion point) - 1
            System.out.println("Index: " + index + " midpoint: " + midpoint);
            index = -index - 1;
            if (index != 0 && index != sortedTokens.length - 1)
            {
                // Check to see which neighbor is closer
                var leftNeighbor = sortedTokens[index - 1];
                var rightNeighbor = sortedTokens[index];
                index = leftNeighbor.size(midpoint) <= midpoint.size(rightNeighbor) ? index - 1 : index;
            }
        }
        var tokenAlignedMidpoint = sortedTokens[index];
        var leftSplitPointCount = splitPointCount / 2;
        var rightSplitPointCount = splitPointCount - leftSplitPointCount;
        if (leftSplitPointCount > 0)
            findTokenAlignedSplitPoints(sortedTokens, min, index, left, tokenAlignedMidpoint, leftSplitPointCount, splitPoints);
        // Add this split point after finding all split points to the left
        splitPoints.add(tokenAlignedMidpoint);
        if (rightSplitPointCount > 0)
            findTokenAlignedSplitPoints(sortedTokens, index + 1, max, tokenAlignedMidpoint, right, rightSplitPointCount, splitPoints);
    }
```

That code doesn't work because there is no guarantee that the sorted tokens are split equally. Is that design close to what you were thinking? Or do I literally need to add calls to remove tokens from `sortedTokens` and then at each successive recursive call I would get the next power of two worth of tokens until I hit the `sortedTokens`.

Reply:
Yes, I did have literal removal from the list in mind. It seems, however, that we can make your original solution work by making sure we always report the right number of splits, by doing something like this:
```
        int pos = 0;
        for (int i = 0; i < evenSplitPoints.length; i++)
        {
            int min = pos;
            int max = sortedTokens.length - evenSplitPoints.length + i;
            Token value = evenSplitPoints[i];
            pos = Arrays.binarySearch(sortedTokens, min, max, value);
            if (pos < 0)
                pos = -pos - 1;

            if (pos == min)
            {
                // No left neighbor, so choose the right neighbor
                nodeAlignedSplitPoints[i] = sortedTokens[pos];
                pos++;
            }
            else if (pos == max)
            {
                // No right neighbor, so choose the left neighbor
                // This also means that for all greater indexes we don't have a choice.
                for (; i < evenSplitPoints.length; ++i)
                    nodeAlignedSplitPoints[i] = sortedTokens[pos++ - 1];
            }
            else
            {
                // Check the neighbors
                Token leftNeighbor = sortedTokens[pos - 1];
                Token rightNeighbor = sortedTokens[pos];

                // Choose the nearest neighbor. By convention, prefer left if value is midpoint, but don't
                // choose the same token twice.
                if (leftNeighbor.size(value) <= value.size(rightNeighbor))
                {
                    nodeAlignedSplitPoints[i] = leftNeighbor;
                    // No need to bump pos because we decremented it to find the right split token.
                }
                else
                {
                    nodeAlignedSplitPoints[i] = rightNeighbor;
                    pos++;
                }
            }
        }
```

Could you run this through your test to see if it works? If not, could you upload the test so that I can play with it?

Reply:
That passed my test, thank you! I'll add some more tests proving that as we add nodes, the split points continue to be aligned

Reply:
This algorithm fails to choose the same tokens if we add one shard at a time. For example, give sorted tokens:

`[-9193912203636246467, -4846620020038852605, -1955575638654777768, 1845313162618248341, 4442481910831405714, 7317158111889931131]`

3 shards chooses split points `[-1955575638654777768, 1845313162618248341]`
4 shards chooses split points `[-4846620020038852605, 1845313162618248341, 4442481910831405714]`

Observe that the 3 split points do not contain 2 split points. I think this is acceptable because the shard count is always `baseShardCount * 2 ^ n`. Is that correct? I'll push up a test shortly that confirms this works with power of 2 growth.

Reply:
I just pushed a commit with `testRangeEndsAreFromTokenListAndContainLowerRangeEnds` to show that this works for powers of 2 shard counts.

Reply:
Yeah, we are only guaranteed that `findNearest(x)` is a subset of `findNearest(y)` when `x` is a subset of `y`. The logic is that if we couldn't select the closest value to a point, it is only because we have already picked it for another point. This shouldn't be too hard to prove formally if we need to do that (I did check if ChatGPT 4o is smart enough to do it, without success).